gobblin-salesforce/src/main/java/gobblin/salesforce/SalesforceSource.java

```diff
@@ -61,6 +93,200 @@
+        // Summarize current group
+        statistics.addValue(count);
+        // A step-in start
+        partitionPoints.add(Utils.toDateTimeFormat(group.getKey(), DAY_FORMAT, Partitioner.WATERMARKTIMEFORMAT));
+        count = group.count;
+      } else {
+        // Add group into current partition
+        count += group.count;
+      }
+
+      if (count >= interval) {
+        // Summarize current group
+        statistics.addValue(count);
+        // A fresh start next time
+        count = 0;
+      }
+    }
+
+    // If the last group is used as the last partition point
+    if (count == 0) {
+      // Exchange the last partition point with global high watermark
+      partitionPoints.set(partitionPoints.size() - 1, Long.toString(expectedHighWatermark));
+    } else {
+      // Summarize last group
+      statistics.addValue(count);
+      // Add global high watermark as last point
+      partitionPoints.add(Long.toString(expectedHighWatermark));
+    }
+
+    log.info("Dynamic partitioning statistics: ");
+    log.info("data: " + Arrays.toString(statistics.getValues()));
+    log.info(statistics.toString());
+    String specifiedPartitions = Joiner.on(",").join(partitionPoints);
+    log.info("Calculated specified partitions: " + specifiedPartitions);
+    return specifiedPartitions;
+  }
+
+  private Histogram getHistogram(String entity, String watermarkColumn, SourceState state,
+      Partition partition) {
+    Histogram histogram = new Histogram();
+
+    Calendar calendar = new GregorianCalendar();
+    Date startDate = Utils.toDate(partition.getLowWatermark(), Partitioner.WATERMARKTIMEFORMAT);
+    calendar.setTime(startDate);
+    int startYear = calendar.get(Calendar.YEAR);
+
+    Date endDate = Utils.toDate(partition.getHighWatermark(), Partitioner.WATERMARKTIMEFORMAT);
+    calendar.setTime(endDate);
+    int endYear = calendar.get(Calendar.YEAR);
+
+    Map<String, String> values = new HashMap<>();
+    values.put("table", entity);
+    values.put("column", watermarkColumn);
+    StrSubstitutor sub = new StrSubstitutor(values);
+
+    SalesforceConnector connector = new SalesforceConnector(state);
+    try {
+      if (!connector.connect()) {
+        throw new RuntimeException("Failed to connect.");
+      }
+    } catch (RestApiConnectionException e) {
+      throw new RuntimeException("Failed to connect.", e);
+    }
+
+    for (int year = startYear; year <= endYear; year++) {
+      values.put("year", Integer.toString(year));
+      String query = sub.replace(DAY_PARTITION_QUERY_TEMPLATE);
+      log.info("Histogram query: " + query);
+
+      try {
+        String soqlQuery = SalesforceExtractor.getSoqlUrl(query);
+        List<Command> commands = RestApiConnector.constructGetCommand(connector.getFullUri(soqlQuery));
+        CommandOutput<?, ?> response = connector.getResponse(commands);
+        histogram.add(parseHistogram(response));
+      } catch (Exception e) {
```

Comment:
Please catch specific exceptions if possible.  Also, shouldn't the exception be propagated instead of suppressed?